contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,1051 @@
+ *
+ * - Increasing the Position (Lending):
+ *   - The lowest buckets with AA are used first
+ *   - Decreases AA
+ *   - Increases OP
+ *
+ * - Decreasing the Position (Being Paid-Back)
+ *   - The highest buckets with OP are paid back first
+ *   - Decreases OP
+ *   - Increases AA
+ *
+ *
+ * - Over time, this gives highest interest rates to earlier buckets, but disallows withdrawals from
+ *   those buckets for a longer period of time.
+ * - Deposits in the same bucket earn the same interest rate.
+ * - Lenders can withdraw their funds at any time if they are not being lent (and are therefore not
+ *   making the maximum interest).
+ * - The highest bucket with OP is always less-than-or-equal-to the lowest bucket with AA
+ */
+contract BucketLender is
+    OnlyMargin,
+    LoanOwner,
+    IncreaseLoanDelegator,
+    MarginCallDelegator,
+    CancelMarginCallDelegator,
+    ForceRecoverCollateralDelegator,
+    LoanOfferingVerifier,
+    ReentrancyGuard
+{
+    using SafeMath for uint256;
+
+    // ============ Events ============
+
+    event Deposit(
+        address beneficiary,
+        uint256 bucket,
+        uint256 amount,
+        uint256 weight
+    );
+
+    event Withdraw(
+        address withdrawer,
+        uint256 bucket,
+        uint256 weight,
+        uint256 owedTokenWithdrawn,
+        uint256 heldTokenWithdrawn
+    );
+
+    // ============ State Variables ============
+
+    /**
+     * Available Amount (AA) is the amount of tokens that is available to be lent by each bucket.
+     * These tokens are also available to be withdrawn by the accounts that have weight in the
+     * bucket.
+     */
+    // AA for each bucket
+    mapping(uint256 => uint256) public availableForBucket;
+    // Total AA
+    uint256 public availableTotal;
+
+    /**
+     * Outstanding Principal (OP) is the share of the margin position's principal that each bucket
+     * is responsible for. That is, each bucket with OP is owed (OP)*E^(RT) owedTokens in repayment.
+     */
+    // OP for each bucket
+    mapping(uint256 => uint256) public principalForBucket;
+    // Total OP
+    uint256 public principalTotal;
+
+    /**
+     * Weight determines an account's proportional share of a bucket. Relative weights have no
+     * meaning if they are not for the same bucket. Likewise, the relative weight of two buckets has
+     * no meaning. However, the relative weight of two accounts within the same bucket is equal to
+     * the accounts' shares in the bucket and are therefore proportional to the payout that they
+     * should expect from withdrawing from that bucket.
+     */
+    // Weight for each account in each bucket
+    mapping(uint256 => mapping(address => uint256)) public weightForBucketForAccount;
+    // Total Weight for each bucket
+    mapping(uint256 => uint256) public weightForBucket;
+
+    /**
+     * The critical bucket is:
+     * - Greater-than-or-equal-to The highest bucket with OP
```

Comment:
Ok I've decided that I find the AA/OP abbreviations to be confusing haha. it's not too much more to just say outstanding principal / available amount, or the variable versions `outstandingPrincipal`/`availableAmount`

It's not a huge deal though